### ISSUE TITLE
Make needless_pass_by_value Allow by default

### ISSUE THREAD

**Issue by birkenfeld** [rust-lang/rust-clippy#2372]
Besides many [false positives](https://github.com/rust-lang-nursery/rust-clippy/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+needless_pass_by_value), `needless_pass_by_value` impedes API design.

It is very possible that I don't need to technically move ownership, but that it makes sense from an API design perspective, to signal "you shouldn't use this argument afterwards".

In any nontrivial codebase, this is bound to occur dozens of times, so I would propose making this lint silent by default.

**Comment by Manishearth** [comment]
I think such APIs are relatively rare and should use allow.

We should try to fix the false positives (I'll take a crack at it). Most
seem to be type based. Of we still have a lot then perhaps we should mark
it as allow by default till stuff is fixed.

**Comment by Manishearth** [comment]
https://github.com/rust-lang-nursery/rust-clippy/pull/2375

**Comment by birkenfeld** [comment]
Hm, just saw this again on `Arc<Mutex<...>>` sent into a thread.

It's really just a useless added indirection to force taking these as references.

**Comment by oli-obk** [comment]
In my opinion: If you are not cloning the `Arc`, you should be passing `&Mutex<T>`. If you are always locking the mutex, you should be passing `&mut T`. And if you are cloning the `Arc` that you passed by value, you're doing it wrong.

Can you give a concrete example?

**Comment by birkenfeld** [comment]
Simplified:
```
        let foo = Arc::new(Mutex::new(some_foo));

        // start a thread that uses it
        let foo_clone = foo.clone();
        thread::spawn(move || Some::thread_function(foo_clone));

        // start another such thread
        let foo_clone = foo.clone();
        thread::spawn(move || Other::thread_functionupdater(r_updates));

        Ok(Server { db, upd_q: w_updates })
```

**Comment by Manishearth** [comment]
I think the correct thing to do there is make the closure explicitly use a local, really.

The problem occurs her because closure captures are guessed from the syntax you use, _and_ because you've decided the function is only to be used when spawning a thread.

I still think it's correct from an API standpoint that those functions should take `&Mutex` -- your additional constraint that these functions should only be called when spawning threads is special enough that it merits an allow -- Clippy is meant to be used with a liberal sprinkling of allow.

**Comment by mitsuhiko** [comment]
This clippy is a major pain for users of actix-web. The extractor system moves values in and that's the API contract. I keep having to disable this clippy lint on every single actix endpoint. Here an example signature:

```rust
#[cfg_attr(feature = "cargo-clippy", allow(needless_pass_by_value))]
fn get_project_configs(
    state: CurrentServiceState,
    body: SignedJson<GetProjectStates>,
) -> ResponseFuture<Json<GetProjectStatesResponse>, Error> {
```

Even though I could get away with a `&CurrentServiceState` that would not compile because then the function signature is incompatible with what actix requires.

**Comment by Manishearth** [comment]
Can you link to a full example? Is this a trait method? We should turn off the lint for traits if we don't already.

**Comment by mitsuhiko** [comment]
This is a full endpoint we have: https://github.com/getsentry/semaphore/blob/master/server/src/endpoints/public_keys.rs

The handler signature from actix-web is this is this:

```rust
    pub fn with<T, F, R>(&mut self, handler: F)
    where
        F: WithFactory<T, S, R> + 'static,
        R: Responder + 'static,
        T: FromRequest<S> + 'static,
    {
        self.h(handler.create());
    }
```

Where `WithFactory` is quite a beast to enable variadic handler functions for dependency injection: https://github.com/actix/actix-web/blob/master/src/with.rs

**Comment by birkenfeld** [comment]
Fixed by #3439
